### ISSUE TITLE
[RFC]: Orthogonal PCP and DCP on vLLM Ascend Model Runner V2

### ISSUE THREAD

**Issue by pisceskkk** [vllm-project/vllm-ascend#12622]
# RFC: Orthogonal PCP and DCP on vLLM Ascend Model Runner V2

Upstream design [vLLM RFC #46358: Unify Context Parallelism in vLLM (PCP ⊥ DCP)](https://github.com/vllm-project/vllm/issues/46358) 

## Summary

This RFC proposes Prefill Context Parallelism (PCP) support for vLLM Ascend on Model Runner V2 and aligns Decode Context Parallelism (DCP) with the upstream orthogonal-groups design.

PCP and DCP solve different long-context serving bottlenecks:

- PCP partitions prefill sequence computation across additional ranks.
- DCP partitions the decode KV cache without adding ranks to the process world.

They must therefore be independent parallel axes. PCP is a compute-partitioning axis and must not increase the KV-cache shard count. DCP is the only owner of the sequence-sharded KV-cache layout:

```text
process world size = TP * PCP * PP * DP
KV-cache shard count = DCP
scheduler block size = physical KV block size * DCP
```

The first Ascend implementation will use the upstream MRV2 virtual-batch PCP manager and initially support the two regular PCP+DCP topologies defined by the upstream RFC:

1. `DCP = PCP`
2. `DCP = TP * PCP`

Arbitrary intermediate DCP layouts are out of scope for the first version.

## Motivation

Long-context prefill is dominated by attention computation and communication, while decode is dominated by KV-cache capacity and bandwidth. Deployments with very long contexts need to split the prefill work and the decode cache at the same time. Treating PCP and DCP as factors of one combined CP dimension causes incorrect ownership semantics, unnecessarily scales scheduler block sizes by
PCP, and complicates composition with TP, sparse attention, graph capture, speculative decoding, and P/D disaggregation.

vLLM Ascend already has Model Runner V1 DCP implementations for GQA, MLA, and SFA, plus a separate DSA-CP path. Recent preparatory changes have removed the legacy PCP coupling from those DCP backends. However, the Ascend MRV2 runner
still rejects both PCP and DCP. This RFC defines the migration from that DCP-only baseline to an upstream-compatible MRV2 implementation.

## Goals

- Reuse the upstream MRV2 PCP virtual-batch lifecycle instead of maintaining an   Ascend-specific scheduler or batch-partitioning implementation.
- Make PCP and DCP orthogonal in process groups, slot mapping, block accounting, attention metadata, and KV transfer.
- Preserve one backend source of truth: Ascend DCP/PCP implementations should specialize normal Ascend attention builders and implementations rather than   duplicate them.
- Support GQA, MLA, SFA, and DSA attention families.
- Support hybrid models that combine full attention with linear/recurrent backends such as GDN/Mamba-style layers.
- Define composition requirements for MTP, ACL Graph, and P/D disaggregation.
- Add correctness and performance gates for every supported topology.

### Parallel axes

The logical rank layout follows upstream vLLM:

```text
ExternalDP × DP × PP × PCP × TP
```

- TP partitions attention heads and tensor-parallel weights.
- PCP partitions prefill query tokens and expands the process world.
- DCP reuses existing ranks and partitions token positions in the decode cache.
- DP remains an independent request-replication axis. PCP+DP support is deferred until upstream removes its current restriction.

### Ownership

The following invariants are mandatory:

1. PCP never changes scheduler block size, hash granularity, or per-rank KV capacity.
2. Slot ownership depends only on `dcp_rank`, `dcp_size`, and the configured interleave size.
3. A PCP-only deployment has one complete KV-cache replica per PCP rank.
4. A PCP+DCP deployment stores one DCP shard per rank. PCP ranks contribute prefill K/V values but do not define an additional storage shard.
5. Recurrent state used by a linear-attention backend is not automatically a DCP-sharded attention KV cache. Each state type must define its own PCP/DCP ownership and merge operation.

### MRV2 virtual-batch lifecycle

The engine core and scheduler keep a global batch. The worker partitions only the per-step MRV2 input:

```text
global InputBatch
    |
    v
PCPManager.partition_batch()
    |
    +-- rank-local input IDs and positions
    +-- rank-local virtual request rows
    +-- rank-local block tables
    +-- expanded slot mappings
    |
    v
Ascend model forward
    |
    v
PCP all-gather + restore index
    |
    v
global token order for logits, sampling, and post-processing
```

For PCP size `P`, each prefill request is divided into `2P` chunks. PCP rank `r` receives chunks `r` and `2P - 1 - r`, which balances causal-attention work. Decode rows are replicated across PCP ranks. Rank-local rows are padded to a common token count so that collectives and graph replay use compatible shapes.

The Ascend runner should inherit this lifecycle from upstream. Ascend-specific code is limited to device buffers, attention metadata, collective kernels, stream ordering, and ACL Graph integration.

### KV-cache insertion

Each PCP rank computes latent or regular K/V only for its local prefill tokens.
Before cache insertion:

- partitioned prefill K/V is all-gathered across the PCP group;
- replicated decode K/V is kept local and written once per local cache;
- slot mappings use DCP ownership only;
- padding and duplicate decode entries are mapped to the padding slot.

Sparse indexer keys follow the same rule. The indexer cache may be replicated or DCP-sharded according to the backend contract, but it must not implicitly use `PCP * DCP` as its shard count.

#### Case A: `DCP = PCP`

Members of a DCP group have the same TP position and therefore the same local Q heads. Each rank computes attention against its local KV shard and returns a partial output and softmax LSE. An LSE-aware all-reduce across the DCP/PCP group produces the complete output for the original local heads. No Q all-gather or head scatter is required.

#### Case B: `DCP = TP * PCP`

The DCP group spans the complete TP×PCP block. Q heads are first all-gathered across TP, each rank computes all heads against its local KV shard, and partial outputs are merged using LSE. The result is then reduced or selected back to the original TP-local head range.

On Ascend, the initial correctness implementation may reuse `torch_npu.npu_attention_update` after HCCL all-to-all/all-gather exchange. A later optimization may replace flat collectives with hierarchical PCP then TP collectives when profiling shows a benefit.

### Backend contract

Each attention backend that advertises PCP support must provide or inherit:

- rank-local metadata construction for virtual request rows;
- prefill K/V or Q exchange for PCP;
- DCP-local sequence lengths and block-table views;
- partial output and LSE production for decode;
- an LSE-aware DCP merge with a documented output head layout;
- graph-safe persistent workspaces where graph mode is supported;
- explicit capability checks for unsupported combinations.

The existing `DCPMetadataBuilderMixin` and `DCPImplMixin` remain the common DCP foundation. PCP batch partitioning and restoration remain in upstream MRV2. Backend-specific Ascend code should not recreate a second PCP manager.

### GQA

GQA will specialize the normal Ascend flash-attention backend.

- PCP prefill consumes rank-local query chunks and gathers the K/V fragments needed by the selected causal-attention strategy.
- `DCP = PCP` keeps local Q heads and performs an LSE-aware output all-reduce.
- `DCP = TP * PCP` gathers Q heads across TP before local-shard attention and restores TP-local output heads afterward.
- Chunked and cached prefill must reuse the same DCP-local cache ownership as decode.

### MLA

MLA follows the upstream latent-cache PCP design.

- All-gather `kv_c_normed` and positional key fragments for prefill cache insertion.
- Keep replicated decode writes local.
- Reuse the Ascend MLA DCP implementation for local-shard attention and LSE merge.
- Define behavior for quantized latent caches and fused output quantization; unsupported fused combinations must fail during initialization.

### SFA

SFA requires two coordinated cache paths: the attention KV cache and the sparse indexer cache.

- Partition/gather indexer keys consistently with MRV2 virtual rows.
- Decide explicitly whether the indexer is replicated or DCP-sharded for each supported model and kernel.
- Build compressed slot mappings from DCP ownership, not PCP×DCP ownership.
- Preserve empty-prefill and short-extend behavior after PCP partitioning.
- Add parity tests for sparse selection indices as well as final model output.

### DSA

DSA-CP is an attention algorithm/data-layout path and is not a synonym for PCP. The existing DSA-CP implementation must be composed with the MRV2 virtual batch:

- virtual-token ordering must be propagated into DSA selection metadata;
- prefill/mixed and decode-only output layouts must remain distinguishable;
- the mixed `o_proj` path must preserve its TP-sharded persistent parameter and use temporary full-weight buffers only when required;
- DSA-CP and DCP collectives must have explicit, non-overlapping group roles.

### Hybrid linear-attention backends

Models such as Qwen3-Next and Qwen3.5 mix full attention with recurrent linear layers such as GDN/Mamba-style backends. Their state cannot use the ordinary attention KV-sharding rule blindly.

- Full-attention groups use DCP-only KV ownership.
- Recurrent state groups define whether state is replicated or passed between PCP ranks during prefill.
- PCP prefill must perform the required prefix-state scan in sequence order, including empty and uneven rank segments.
- Scheduler and hybrid cache managers must support a global request whose attention cache is DCP-sharded while recurrent state has a different layout.
- Mixed batches, prefix caching, preemption, and state zeroing need dedicated tests.

### MTP speculative decoding

The upstream MRV2 PCP baseline currently rejects speculative decoding. Ascend must extend both target and draft paths before enabling PCP+MTP:

- partition base and draft token rows without changing acceptance semantics;
- generate per-step attention metadata for the rank-local virtual batch;
- keep DCP MTP masks and local sequence lengths aligned with DCP-only KV ownership;
- restore target hidden states and logits before rejection sampling;
- support sync and async scheduling without stale PCP restore metadata;
- cover one-step and multi-step MTP, mixed prefill/decode, and graph replay.

### ACL Graph mode

PCP introduces dynamic rank-local request counts and uneven token counts. Graph capture therefore requires persistent padded buffers and topology-aware graph keys.

- Start with eager correctness.
- Enable piecewise ACL Graph for dense GQA/MLA after all collectives and slot mappings are capture-safe.
- Enable full graph only after input partition, metadata update, cache insertion, and output restoration have stable addresses and replay-safe stream ordering.
- SFA/DSA and hybrid linear backends require separate capture qualification.
- Graph keys must distinguish PCP/DCP sizes, backend layout, MTP query length, and mixed-batch state.

### P/D disaggregation

P/D disaggregation must transfer storage shards, not PCP compute partitions. Existing code that uses `pcp_size * dcp_size` for block scaling, rank namespaces, or transfer planning must be migrated.

- Advertise TP, PCP, DCP, PP, KV group, and cache-layout metadata separately.
- Compute transferable KV shards from DCP only.
- Treat PCP-only ranks as cache replicas and select or deduplicate sources.
- Support different P-side and D-side TP/PCP/DCP layouts with an explicit source-to-destination shard map.
- Transfer SFA indexer caches and hybrid recurrent states according to their own layout contracts.
- Preserve prefix-cache boundaries, compressed-cache strides, and MTP cache groups.
- Add reconnect, abort, partial-prefix, and multi-node tests.

## Compatibility and failure policy

Unsupported combinations must fail during configuration or model-runner initialization. They must not fall back silently to a combined `PCP * DCP` layout. The first release should retain the upstream restrictions on PCP+PP, multimodal inputs, LoRA, and unsupported attention backends unless the corresponding feature has an explicit owner and test plan.

Model Runner V1 remains DCP-only. Setting `prefill_context_parallel_size > 1` with MRV1 continues to fail.

## Implementation checklist and MRV2 adaptation progress

The following status reflects the tracking baseline at the top of this RFC. Checked items are preparatory work already present; they do not mean that MRV2 PCP is user-ready on Ascend.

- [x] Remove PCP from the Ascend MRV1. https://github.com/vllm-project/vllm-ascend/pull/12592
- [x] MRV2 Basic Support https://github.com/vllm-project/vllm-ascend/pull/12595
- [x] GQA https://github.com/vllm-project/vllm-ascend/pull/14023
- [x] MLA https://github.com/vllm-project/vllm-ascend/pull/14008
- [ ] SFA https://github.com/vllm-project/vllm-ascend/pull/14026/
- [x] DSA https://github.com/vllm-project/vllm-ascend/pull/14037
- [ ] Hybrid linear-attention backends 
- [ ] Sepc Decoding
- [ ] ACL Graph mode
- [ ] P/D disaggregation

## References

- [vLLM RFC #46358: Unify Context Parallelism in vLLM (PCP ⊥ DCP)](https://github.com/vllm-project/vllm/issues/46358)
- [vLLM MRV2 virtual-batch PCP implementation](https://github.com/vllm-project/vllm/commit/866c998c033f422b3cfe903f47b96e1cd2cfc905)
